### ISSUE TITLE
Out-of-distribution generalization

### ISSUE THREAD

**Issue by Expertium** [open-spaced-repetition/srs-benchmark#166]
Some time ago I plotted this for FSRS-4.5:

https://i.imgur.com/DSjaW5e.png

https://i.imgur.com/1JsW2Jy.png

(can't upload images directly for some reason)

While the exact numbers will vary depending on the dataset and the outlier filter and whatnot, the trend is clear: lower decay is better.

...as long as retention stays high.

https://discord.com/channels/368267295601983490/1282005522513530952/1332619385440964670

> If I use the simulated data to optimize FSRS with trainable forgetting decay, the decay will never approach 0.5 which is the default value of forgetting decay and used to generate the simulated data. 
> The optimizer prefers flat forgetting curves and I don't know the reason.

https://discord.com/channels/368267295601983490/1282005522513530952/1332670300013199411

> New Finding: the decay increases when I decreases the desired retention of the simulation config.
> So the decay is highly relevant to retention of training data.

If true R is always >50%, then there is no reason for the algorithm to learn to predict R lower than 50%. So the "optimal" value of decay depends on the training data. And most Anki users have retention around 80-95%, without too many insanely overdue cards. This means that any attempt to fit the curve better to the dataset is bound to cause issues in practice when the algorithm will vastly overestimate R for overdue cards, or worse: for cards that are scheduled with low DR.

# If someone had average retention of 95%, then switched to FSRS and is now using desired retention=70%, they are screwed.

More generally speaking, what shape of the curve is optimal depends on retention.

So, what do we do about it?

@L-M-Sherlock @1DWalker 

EDIT: to avoid confusion, let me clarify - I believe this is a problem both on the benchmark level and on the level of individual users. On the benchmark level, the "optimal" value of decay depends on the average retention across all collections. On an individual level I suspect that even with a fixed value of decay FSRS still learns to never predict low values of R by predicting very large S. This matches my experience with r/Anki posts, where even people with good metrics, like 2-3% RMSE, report >5% discrepancy between desired retention and true retention. The only explanation is that **FSRS doesn't generalize well between different levels of retention**.

**Comment by Expertium** [comment]
I forgot @user1823 

IMO this is a serious problem. "If we don't solve it, we're f###ed" kind of serious. I'd like to have as many people pay attention to it as possible.

**Comment by Expertium** [comment]
My idea that I shared on Discord:

How about this:
1) Calculate retention within a given batch, let's call it `retention_batch`.
2) Use an exponential moving average, like this: `retention_ma = retention_batch * a + retention_ma * (1-a)`, where `retention_ma` is the moving average that gets updated after every batch. For example, if retention in a new batch was 0.8, `retention_ma` was 0.9, and a = 0.5, new value of `retention_ma` would be 0.85. `a` is a hyper-parameter that needs to be fine-tuned.
3) Adjust the loss like this: `loss = loss * abs(retention_ma_new - retention_ma_prev) * b`, where `retention_ma_new` is the new value of the moving average after it was updated on the current batch, and `retention_ma_prev` is the previous value of the moving average. `b` is a hyperparameter that needs to be fine-tuned. If necessary, the abs() term can be clamped to some non-zero value, to ensure that the loss is never too close to 0.

The idea is to update parameters harder if there is drift in retention. If the difference between the latest and the previous values of the moving average is large, that means that retention is changing and we need to update parameters really hard. If the difference is small, that means that retention is stable, and we don't need to update too hard.

This is a more adaptive alternative to recency weighting.

A slightly different approach proposed by 1DWalker: 

> Take v for velocity, update it using exponential moving average based on abs(retention_ma_new - retention_ma_prev)^t. Then we have loss = loss * (1 + sv) or something
> s is a scaling factor

This is *a* solution, but it treats the symptoms, not the underlying problem. It would just make FSRS oscillate between being good at one level of retention and being good at another level of retention. What we actually want is for FSRS to be good at ***ALL*** levels of retention simultaneously.

**Comment by Expertium** [comment]
Another idea: we could ask Dae for a new dataset (again), but this time ask him to calculate retention for each user and make sure the dataset has the following properties:
1) The average retention across all users is roughly 50%
2) The number of users with retention higher than 50% is roughly the same as the number of users with retention lower than that

I'm sure among millions of Anki users it's possible to find 5k people with <50% retention.
Then we will be able to fine-tune the shape of the forgetting curve without having to worry about overfitting to people with high retention. We would still have to worry about lack of generalziation on the individual level, but at least this would be a step forward.

Even better: make a dataset with uniformly distributed retention, such that any value of retention is equally likely to appear in the dataset. Idk if Dae would want to put in the work to do that, and since he's the only one who has access to all that data, nobody can so it for him.

**Comment by JSchoreels** [comment]
Anecdotally encounter that issue when I used the "Compute my Optimal Retention", which recommended me to drop from .80 DR to .70. The problem was my 5-day average .80 became .60 very quickly, and on a specific day it dropped as low as ~52%, at this point I backed up to .80 to avoid burying my own grave.

My gut feeling was and still is that unfortunately, while FSRS can be good at predicting what it has been trained to predict, the whole Forgetting Curve might not really fit how a prediction for .80 DR can be mapped to a .70 DR. 

I also think that until this is at least under control, any kind of "Automatic Desired Retention" should be avoided. FSRS is able to predict a certain DR, but is far from being able to predict others.

**Comment by user1823** [comment]
I really don't know how to test this but maybe we need to go back to the exponential forgetting curve. Probably, the [better metrics that we saw with a power curve](https://github.com/open-spaced-repetition/fsrs4anki/issues/282#issuecomment-1580670961) were just due to this bias in the evaluation.

For those who are unaware, FSRS v3 used an exponential forgetting curve, switched to power curve (with -1 decay) in FSRS v4 and switched to -0.5 decay in FSRS v4.5

Also see:
- https://github.com/open-spaced-repetition/fsrs4anki/issues/274

**Comment by Expertium** [comment]
That would just make the metrics worse without really telling us much otherwise. I think we need a new dataset as described here: https://github.com/open-spaced-repetition/srs-benchmark/issues/166#issuecomment-2614417100

Then we can test whether the exponential curve is better on _that_ dataset.

**Comment by Expertium** [comment]
@1DWalker about this: https://github.com/open-spaced-repetition/srs-benchmark/issues/166#issuecomment-2614160163

On second thought, I'm not sure if this is a good idea. It will just make FSRS oscillate between "Good at retention=x% and bad at retention=y%" and "Good at retention=y% and bad at retention=x%". The user's retention drops to 70% - FSRS adapts to 70% retention. The user's retention increases to 95% - FSRS adapts to 95% retention. We need some way to make FSRS good at all levels, but idk how.

P.S. Regarding the new uniform(ish) dataset, I still need a bit more time. It should be finished tomorrow. Then I'll try different valeus of decay (and the exponential curve, too). That will at least give us some assurance that the shape of the curve is good. But we still need some method to improve generalization across different retention levels.

**Comment by Expertium** [comment]
Well, nevermind, this new dataset isn't nearly uniform enough despite my best efforts

Users=1000
10th percentile=46.4%
20th percentile=52.6%
30th percentile=56.3%
40th percentile=58.8%
50th percentile=61.2%
60th percentile=63.8%
70th percentile=69.9%
80th percentile=80.0%
90th percentile=90.0%

For comparison, here are the stats for the full 10k dataset
10th percentile=67.6%
20th percentile=74.7%
30th percentile=79.4%
40th percentile=82.9%
50th percentile=85.8%
60th percentile=88.6%
70th percentile=91.4%
80th percentile=94.1%
90th percentile=96.9%

I guess we'll have to ask Dae to make a 1mil dataset, and then I can "extract" a uniform 10k dataset from it. Or at least a 100k dataset so that I can "extract" a 1k uniform dataset from it.
@L-M-Sherlock

**Comment by Expertium** [comment]
Alright, I've made a very small 100 users dataset where retentions are distributed quite uniformly across users:
10th percentile=23.2%
20th percentile=31.9%
30th percentile=35.3%
40th percentile=40.0%
50th percentile=50.0%
60th percentile=60.0%
70th percentile=70.0%
80th percentile=80.0%
90th percentile=90.0%
You can download the dataset here: https://drive.google.com/file/d/1iHq26LpDOlCWBSHeymjy6koSWnaodBrR/view?usp=sharing

I've tested many forgetting curves on it, here are the results (I disabled L2 regularization and recency weighting btw):

![Image](https://github.com/user-attachments/assets/cae0d140-6f43-4d35-81e9-9455ac4efe1d)

![Image](https://github.com/user-attachments/assets/447057dd-3502-41b9-8630-630167f2cb18)

![Image](https://github.com/user-attachments/assets/76527f09-c2ec-4234-bb0b-0ed7f0e8a889)

As you can see, the exponential curve is completely awful; for the power curve decay=-0.4 provides the best fit in terms of RMSE and decay=-0.3 provides the best fit in terms of logloss. Of course, ideally, we should test this on much more data, but this is all I've got. The larger I want the uniform dataset to be, the harder it is to select users in a way that makes retentions uniform.

The metrics are worse than on the whole dataset, but the optimal value of decay is still definitely in the "very flat" territory and the current absolute value of decay (0.5) is still greater than optimal.

On the whole dataset, the optimal value of decay is around -0.15 to -0.2. I was expecting the optimal value of decay on this dataset to be **_VERY_** different, maybe even -1.0, but nope, it's only somewhat different. This means that the problem with generalization is not due to the shape of the curve, but due to something else.

@L-M-Sherlock @user1823 @1DWalker sorry for frequent pings, but this is important.

Here's how we can address the problem with generalization: weighting lapses. I know Jarrett has been opposed to that idea before, but hear me out. Without assigning more weight to lapses, we can't prevent algorithms from overfitting to high retentions due to the imbalance in the number of successful reviews and lapses, since in practice the former vastly outnumber the latter. So how exactly should lapses be weighted? Simply: w(label=1)=1, w(label=0)=n_successes/n_lapses.

Example 1: there are 20 reviews, 10 successes and 10 lapses. w(label=1)=1, w(label=0)=10/10=1. The sum of weights of successes=1⋅10=10, the sum of weights of lapses=1⋅10=10.
Example 2: there are 100 reviews, 99 successes and 1 lapse. w(label=1)=1, w(label=0)=99/1=99. The sum of weights of successes=99⋅1=99, the sum of weights of lapses=1⋅99=99.

*Note: in practice I recommend using (n_successes+epsilon)/(n_lapses+epsilon), to avoid zero weights and division by 0; where epsilon can be, for example, 1e-5.

As you can see, this makes it so that regadless of retention, the weights of both types of reviews have the same sum. Loosely speaking, this means that **the algorithm pays the same amount of attention to both successful reviews and lapses**, meaning that being overly optimistic and always predicting high R is not a viable option.
Of course, this weighting scheme must be used both in training and in evaluation. This means re-running the entire benchmark. I know this isn't fun, but it's a small price to pay if we will get a better FSRS as a result. We can add a new section with new tables.

We won't be able to have a direct apples-to-apples comparison between FSRS with and without my "equal sum" weighting, so I'm not sure how we would judge whether it's a success (other than tracking the number of complaints about mismatch between desired and true retention). Hence why I'm asking you guys.

There is another issue: applying weights when calculating the logloss is straightforward, but idk about AUC. And RMSE (bins) is definitely not compatible with this weighted approach.

What do you guys think?

TLDR: the curve is not the problem, something else is. Let's weigh lapses proportionally more so that lapses and successful reviews are treated equally.

**Comment by 1DWalker** [comment]
Be careful, weighting lapses differently than successes can change the value that log loss optimizes for. In your example, suppose we are just training 1 value that predicts the probability of success. We have 100 reviews, 99 successes, and 1 lapse. The loss function is f(p) = -(99*log(p) + log(1-p)) with a minimum value at p=0.99. But if you weigh the lapses by 99x, the loss function is now f(p) = -(99log(p) + 99log(1-p)) with a minimum at p=0.50 which is wrong.

**Comment by L-M-Sherlock** [comment]
I did more experiments since last week. I found that the optimizer prefers flatter forgetting curve with lower stability:

<img width="721" alt="Image" src="https://github.com/user-attachments/assets/f3a91047-27dd-4ae0-ae30-f33aa64f2684" />

https://github.com/open-spaced-repetition/fsrs-trained-on-simulated-data/blob/main/trainable-forgetting-decay.ipynb

The difference is very minor because the fuzz range is too narrow.

**Comment by L-M-Sherlock** [comment]
The problem could be alleviated via introducing more diverse data:

![Image](https://github.com/user-attachments/assets/e24f9574-527d-483e-b9c0-1651d4b7d9b0)

But the decay is still not equal to -1.

**Comment by L-M-Sherlock** [comment]
Finally, I combine three sets of simulated data with DR=0.9, 0.7 and 0.5, the decay is approaching -1.

![Image](https://github.com/user-attachments/assets/1665e644-da9b-4edf-9733-0e08bd1ac424)

**Comment by Expertium** [comment]
Huh, interesting. Have you played around with this dataset?

https://github.com/open-spaced-repetition/srs-benchmark/issues/166#issuecomment-2622748214

**Comment by L-M-Sherlock** [comment]
> Huh, interesting. Have you played around with this dataset?

How to play around with it? Would you mind share your code?

**Comment by Expertium** [comment]
I meant just running algorithms on it

**Comment by L-M-Sherlock** [comment]
OK, I will benchmark https://github.com/open-spaced-repetition/fsrs-optimizer/tree/Expt/trainable-forgetting-decay on them.

**Comment by L-M-Sherlock** [comment]
Here is the result:

<img width="1037" alt="Image" src="https://github.com/user-attachments/assets/5a83c6b1-e7cb-4e8b-9ae0-7dba4bce2082" />

[FSRS-5-dev.txt](https://github.com/user-attachments/files/18763024/FSRS-5-dev.txt)

**Comment by Expertium** [comment]
So yes, flatter curves are more optimal at higher retentions.
I'd say let's keep the current value. We *could* run the optimizer multiple times, each time with a fixed value of decay, but that would make it much slower, and potentially make the problem even worse.

**Comment by Expertium** [comment]
@L-M-Sherlock can you share the numbers from the graph here?
https://github.com/open-spaced-repetition/srs-benchmark/issues/166#issuecomment-2652751232

**Comment by L-M-Sherlock** [comment]
> FSRS-5-dev.txt

Is it?

**Comment by Expertium** [comment]
I don't see decay there

**Comment by L-M-Sherlock** [comment]
It's the last param.
